### ISSUE TITLE
Incorrect results when involving .detach() with second-order derivatives

### ISSUE THREAD

**Issue by mujjingun** [pytorch/functorch#353]
Here's my code that uses the 'straight-through estimator' in the function `q(x)`. In JAX, I would have implemented it using `jax.stop_gradient()` but I don't see that option here, so I used `Tensor.detach()`. If there is another option, please let me know.

```python
import torch
import functorch as ft
import numpy as np

def q(x):
    N = 1
    fake_func = torch.tanh(x)
    return fake_func + (
        torch.clip(torch.round(x), -N, N) - fake_func
    ).detach()

def f_L(x, y, M, lamda):
    return torch.norm(M @ q(y) - M @ x, 2) + lamda * torch.norm(y, 1)

f_y = ft.grad(f_L, argnums=1)
f_yy = ft.jacrev(f_y, argnums=1)

# Test the functions
M = [[-1.7055013,  0.214439 ],
     [ 1.5994732,  1.179864 ]]
x = [ 0.279999 , -1.8200971]
y = [-0.33679163, -1.0274953 ]

M = torch.tensor(np.array(M))
lamda = 0.1

x = torch.tensor(np.array(x))
y = torch.tensor(np.array(y))

print('f_L')
print(f_L(x, y, M, lamda))

print('f_y')
print(f_y(x, y, M, lamda))

print('f_yy')
print(f_yy(x, y, M, lamda))
```

The above code prints
```
f_L
tensor(0.9713, dtype=torch.float64)
f_y
tensor([-0.4033,  0.2633], dtype=torch.float64)
f_yy
tensor([[-0.1969,  0.0000],
        [ 0.0000,  0.5616]], dtype=torch.float64)
```

However, the equivalent code in JAX prints

```
f_L
0.9713378
f_y
[-0.40328947  0.26329258]
f_yy
[[4.933743   0.7883494 ]
 [0.78834945 0.6827129 ]]
```

which is clearly different in the values of `f_yy`.

Btw, thank you for the amazing work! This library is truly awesome.

**Comment by vfdev-5** [comment]
@mujjingun I can't say why it does not work with detach but `torch.no_grad` should work:
```python
def q(x):
    N = 1
    fake_func = torch.tanh(x)
    with torch.no_grad():
        resid = torch.clip(torch.round(x), -N, N) - fake_func
    output = fake_func + resid
    return output
```
Executing the complete code I get:
```python
f_yy
tensor([[4.9337, 0.7883],
        [0.7883, 0.6827]], dtype=torch.float64)
```

**Comment by mujjingun** [comment]
Thank you!

**Comment by zou3519** [comment]
Thanks for the bug report, @mujjingun. This is definitely a bug and as @vfdev-5 suggested, `torch.no_grad` is a workaround for now. We'll look into fixing this

**Comment by zou3519** [comment]
This should be fixed by https://github.com/pytorch/pytorch/pull/71707 ! Please do let us know if it isn't.
